# 문제 발생에 대비하기

## 5.2 항상 가장 구체적인 예외 잡기

```java
class TransmissionParser {
    TransmissionParser parse(String rawMessage) {
        if (rawMessage == null || rawMessage.length() != Transmission.MESSAGE_LENGTH) {
            throw new IllegalArgumentException("Bad message received");
        }

        String rawId = rawMessage.substring(0, Transmission.ID_LENGTH);
        String rawContent = rawMEssage.substring(Transmission.ID_LENGTH);

        try {
            int id = Integer.parseInt(rawId);
            String content = rawContent.trim();
            return new Transmission(id, content);
        } catch (Exception e) {
            throw new IllegalArgumentException("Bad message received");
        }
    }
}
```

예외는 항상 가장 구체적인 예외 타입을 잡아야 한다. 일반적인 타입으로 잡으면 잡아서 안 될 오류까지 잡힐 위험이 있다.

명령문 하나로 모든 실제 오류 유형을 잡을 수 있다는 생각에 일반적인 타입을 잡으려고 한다.
이러한 예외 중 99%는 반드시 수정해야 할 코드 내 버그이다. 문제를 알려면 프로그램이 충동해야 한다.

```java
class TransmissionParser {
    TransmissionParser parse(String rawMessage) {
        if (rawMessage == null || rawMessage.length() != Transmission.MESSAGE_LENGTH) {
            throw new IllegalArgumentException("Bad message received");
        }

        String rawId = rawMessage.substring(0, Transmission.ID_LENGTH);
        String rawContent = rawMEssage.substring(Transmission.ID_LENGTH);

        try {
            int id = Integer.parseInt(rawId);
            String content = rawContent.trim();
            return new Transmission(id, content);
        } catch (NumberFormatException e) { // 변경됨
            throw new IllegalArgumentException("Bad message received");
        }
    }
}
```
예제에서 처리해야 할 구체적인 예외 유형은 `NumberFormatException` 하나 뿐이다.

그러니 `Exception`을 `NumberFormatException`으로 바꾸기만 하면 된다.

catch 블록이 여러 개로 늘어나도 일반적인 예외 유형으로 잡는 것보다 좋다. 버그가 적은 긴 코드가 낫다.


## 5.3 메시지로 원인 설명

```java
class TransmissionParser {
    TransmissionParser parse(String rawMessage) {
        if (rawMessage == null || rawMessage.length() != Transmission.MESSAGE_LENGTH) {
            throw new IllegalArgumentException("Bad message received");
        }

        String rawId = rawMessage.substring(0, Transmission.ID_LENGTH);
        String rawContent = rawMEssage.substring(Transmission.ID_LENGTH);

        try {
            int id = Integer.parseInt(rawId);
            String content = rawContent.trim();
            return new Transmission(id, content);
        } catch (NumberFormatException e) { // 변경됨
            throw new IllegalArgumentException("Bad message received");
        }
    }
}
```

고칠 수 있는 버그는 재현할 수 있는 버그 뿐이다. 입력을 넣지만 실제로 도움이 되지 않는다.

```java

class TransmissionParser {
    TransmissionParser parse(String rawMessage) {
        if (rawMessage == null || rawMessage.length() != Transmission.MESSAGE_LENGTH) {
            throw new IllegalArgumentException(
                    String.format("Expected %d, but got %d characters in '%s'", 
                            Transmisstion.MESSAGE_LENGTH, 
                            rawMessage.length(), 
                            rawMessage));
        }
        
        String rawId = rawMessage.substring(0, Transmission.ID_LENGTH);
        String rawContent = rawMEssage.substring(Transmission.ID_LENGTH);

        try {
            int id = Integer.parseInt(rawId);
            String content = rawContent.trim();
            return new Transmission(id, content);
        } catch (NumberFormatException e) { // 변경됨
            throw new IllegalArgumentException("Expected number, but got '%s' in '%s'", rawId, rawMessage);
        }
    }
}
```
쓸모없는 정보 대신 바라는 것, 받은 것, 전체 맥락 세 가지를 제공한다.

## 5.7 항상 자원 닫기

```java
class Logbook {
    // ...
    List<Path> getLogs() throws IOException {
        List<Path> result = new ArrayList<>();
        DirectoryStream<Path> directoryStream = Files.newDirectoryStream("Path");
        // ... 
        directoryStream.close();
        return result;
    }
}
```

프로그램에는 디스크 공간이나 데이터베이스, 네트워크 연결, CPU 스레드, RAM과 같은 시스템 자원이 필요하다.
이러한 자원은 제한되어 있기 때문에 프로그램은 자원을 서로 공유해야 한다.

더 이상 자원이 필요없으면 바로 해제해야 한다.

위 코드는 `close()` 메서드를 통해 문제가 없어보이지만 해제하기 전에 예외가 발생하면 `close()`가 실행되지 않는다.

```java
import java.nio.file.DirectoryStream;

class Logbook {
    // ...
    List<Path> getLogs() throws IOException {
        List<Path> result = new ArrayList<>();

        try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream("Path")) {
            //...
        }
        return result;
    }
}
```

자바7부터는 try-with-resources 구문으로 자원을 안전하게 닫을 수 있다. `AutoCloseable` 인터페이스를 구현한 클래스여야 동작한다.

```java
DirectoryStream<Path> resource = Files.newDirectoryStream("Path");
try {
    // ...
} finally {
    if (resource != null) {
        resource.close();
    }
}
```
컴파일러가 위와 같이 확장하기 때문에 `NullPointerException`도 피할 수 있다.

> 여러 개 자원을 닫기 위해서는 아래와 같이 사용하면 된다. 
> ```java
> try (open resource1; open resource2) { ... }
> ```
